### PR TITLE
Jms connector Improvements

### PR DESCRIPTION
**Pull request by akalankapagoda**
This pull contains following improvements to the JMS connectors.

1. Restructured how JMS connectors are initialized
2. Added getProperty and setProperty support for messages.
3. Modified how JMS map messages are passed in.

### CODE REVIEW

File: modules/ballerina-native/pom.xml [hunk 1]
```diff
@@ -47,6 +47,10 @@
             <artifactId>HikariCP</artifactId>
         </dependency>
         <dependency>
+            <groupId>org.wso2.carbon.transport</groupId>
```
Comment (isudana) [inline modules/ballerina-native/pom.xml]: Why do we need to depend on jms transport ?

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/connectors/jms/utils/JMSConstants.java [hunk 2]
```diff
@@ -1,84 +0,0 @@
-/*
```
Comment (isudana) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/connectors/jms/utils/JMSConstants.java]: Why we have removed these constants ? I think this made ballerina to directly depend on jms transport

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/messages/GetProperty.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
+/*
```
Comment (isudana) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/messages/GetProperty.java]: @sameera-jayasoma @hasithaa Can you please review this new function ?

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/messages/SetMapPayload.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import org.ballerinalang.natives.AbstractNativeFunction;
+import org.ballerinalang.natives.annotations.Argument;
+import org.ballerinalang.natives.annotations.Attribute;
+import org.ballerinalang.natives.annotations.BallerinaAnnotation;
+import org.ballerinalang.natives.annotations.BallerinaFunction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Native function to get payload as String..
```
Comment (isudana) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/messages/SetMapPayload.java]: Wrong javadoc comment

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/messages/SetProperty.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        value = "The value of the property") })
+public class SetProperty extends AbstractNativeFunction {
+
+    @Override
+    public BValue[] execute(Context context) {
+        BMessage msg = (BMessage) getArgument(context, 0);
+        BString propertyName = (BString) getArgument(context, 1);
+        BString propertyValue = (BString) getArgument(context, 2);
+
+        msg.setProperty(propertyName.stringValue(), propertyValue.stringValue());
```
Comment (isudana) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/messages/SetProperty.java]: Do we need to check for NPEs ?

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/utils/Constants.java [hunk 6]
```diff
@@ -40,4 +40,6 @@
      * HTTP content-type text/plain.
      */
     public static final String TEXT_PLAIN = "text/plain";
+
+    public static final String MAP_TYPE = "application/x-www-form-urlencoded";
```
Comment (isudana) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/utils/Constants.java]: Why map type is equal to form-url-encoded ?

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/net/jms/Commit.java [hunk 7]
```diff
@@ -32,6 +31,9 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.wso2.carbon.messaging.CarbonMessage;
+import org.wso2.carbon.transport.jms.utils.JMSConstants;
```
Comment (isudana) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/net/jms/Commit.java]: Why do we need to depend on constants in the jms transport implementation ? can we get rid of it ?

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/messages/SetMapPayload.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
+/*
+ *  Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
```
Comment (anupama-pathirage) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/messages/SetMapPayload.java]: Copyright year should be 2017.

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/connectors/jms/utils/JMSConstants.java (hunk 2 above)
Comment (akalankapagoda) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/connectors/jms/utils/JMSConstants.java]: Was thinking of removing duplicate constants. As it is better to make Ballerina independent of transport we use, will add them again. Thanks for pointing this out.

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/messages/SetMapPayload.java (hunk 8 above)
Comment (akalankapagoda) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/messages/SetMapPayload.java]: Thanks, will update.

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/messages/SetMapPayload.java (hunk 4 above)
Comment (akalankapagoda) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/messages/SetMapPayload.java]: Thanks, will update.

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/connectors/jms/client/Send.java [hunk 9]
```diff
@@ -90,18 +82,22 @@ public BValue execute(Context context) {
...
             throw new BallerinaException("Ballerina message not found", context);
         }
         CarbonMessage message = bMessage.value();
         //Create property map to send to transport.
         Map<String, String> propertyMap = new HashMap<>();
         //Getting the map of properties.
-        BMap properties = (BMap) getArgument(context, 6);
+        BMap properties = ((ClientConnector) connector).getProperties();
+
+        for (BString key : (Set<BString>) properties.keySet()) {
```
Comment (ramith) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/connectors/jms/client/Send.java]: out of curiosity , can't we iterate entry set?

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/messages/SetMapPayload.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+    @Override
+    public BValue[] execute(Context context) {
+        BMessage msg = (BMessage) getArgument(context, 0);
+        BMap payload = (BMap) getArgument(context, 1);
+        // Clone the message without content
+        msg.setMessageDataSource(payload);
+        msg.setHeader(Constants.CONTENT_TYPE, Constants.MAP_TYPE);
+        if (log.isDebugEnabled()) {
+            log.debug("Setting new map payload");
```
Comment (ramith) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/messages/SetMapPayload.java]: out of curiosity: this debug statement doesn't print anything specific to distinguish a particular method invocation to another. is that intentional ?

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/utils/Constants.java (hunk 6 above)
Comment (akalankapagoda) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/utils/Constants.java]: This is the HTTP content type that maps to the 'Map' data type of JMS. This is generally accepted as the correct type mapping for JMS Map to HTTP Content type and is the case in WSO2 MB as well as IBM MQ.

https://www.ibm.com/support/knowledgecenter/SSFKSJ_9.0.0/com.ibm.mq.ref.dev.doc/q111120_.htm

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/net/jms/Commit.java (hunk 7 above)
Comment (akalankapagoda) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/net/jms/Commit.java]: Yes, will do.

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/connectors/jms/client/Send.java (hunk 9 above)
Comment (akalankapagoda) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/connectors/jms/client/Send.java]: The variable 'properties' in this case is a BMap, which doesn't provide a way to get entry set. Anyway will replace this with stream.collect.

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/messages/SetMapPayload.java (hunk 10 above)
Comment (akalankapagoda) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/messages/SetMapPayload.java]: Not intentional 😄 Will update the debug log to have more specific data.

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/messages/SetProperty.java (hunk 5 above)
Comment (akalankapagoda) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/lang/messages/SetProperty.java]: These are passed in from the ballerina program and is already validated.

File: modules/ballerina-native/pom.xml (hunk 1 above)
Comment (akalankapagoda) [inline modules/ballerina-native/pom.xml]: Yes, we don't need to as explain in the below comment.

**Pull request merged**